### ISSUE TITLE
Broadcasting or-or incurs allocations

### ISSUE THREAD

**Issue by charleskawczynski** [CliMA/ClimaCore.jl#1355]
MWE:

```julia
using Revise, Test
import ClimaCore
if !(@isdefined TU)
    include(joinpath(pkgdir(ClimaCore), "test", "TestUtilities", "TestUtilities.jl"))
    import .TestUtilities as TU
end
foo(a, c) = a;
FT_broadcast_allocating!(a) = FT_broadcast_allocating!(a, eltype(a));
function FT_broadcast_allocating!(a, ::Type{FT}) where FT
    @. a = foo(a, false || a < FT(1))
    # @. a = foo(a, a < FT(1)) # unexpected pass
    return nothing
end;
space = TU.CenterExtrudedFiniteDifferenceSpace(Float32);
a = ClimaCore.Fields.level(fill(Float32(0), space), 1);

FT_broadcast_allocating!(a) # compile first

p_allocated = @allocated FT_broadcast_allocating!(a)
@test_broken p_allocated == 0
```

**Comment by simonbyrne** [comment]
Does broadcasting over `||` actually work?

**Comment by charleskawczynski** [comment]
> Does broadcasting over `||` actually work?

It at least doesn't break

**Comment by charleskawczynski** [comment]
Interestingly,

```julia
using Test
import ClimaCore
import ClimaCore.Fields as Fields
include(joinpath(pkgdir(ClimaCore), "test", "TestUtilities", "TestUtilities.jl"))
import .TestUtilities as TU

function ifelse_broadcast_allocating(a, b, c)
    @. a = ifelse(true || c < b * 1, 0, c)
    return nothing
end

space = TU.CenterExtrudedFiniteDifferenceSpace(Float32);
a = Fields.level(fill(Float32(0), space), 1);
b = Fields.level(fill(Float32(2), space), 1);
c = Fields.level(fill(Float32(3), space), 1);

ifelse_broadcast_allocating(a, b, c)
p_allocated = @allocated ifelse_broadcast_allocating(a, b, c)
@test p_allocated == 0
```
passes. So, it's somehow a combination of `||` and `FT`

**Comment by charleskawczynski** [comment]
It seems unrelated to `ifelse`:

```julia
using Revise
using Test
import ClimaCore
import ClimaCore.Fields as Fields
if !(@isdefined TU)
    include(joinpath(pkgdir(ClimaCore), "test", "TestUtilities", "TestUtilities.jl"))
    import .TestUtilities as TU
end

foo(c, a, b) = a

FT_broadcast_allocating!(a, b, c) =
    FT_broadcast_allocating!(a, b, c, eltype(a))
function FT_broadcast_allocating!(a, b, c, ::Type{FT}) where FT
    @. a = foo(false || c < FT(1), 0, 1) # test is broken
    # @. a = foo(c < FT(1), 0, 1) # unexpected pass
    return nothing
end

space = TU.CenterExtrudedFiniteDifferenceSpace(Float32);
a = Fields.level(fill(Float32(0), space), 1);
b = Fields.level(fill(Float32(2), space), 1);
c = Fields.level(fill(Float32(3), space), 1);

FT_broadcast_allocating!(a, b, c) # compile first (after recompilation)

p_allocated = @allocated FT_broadcast_allocating!(a, b, c)
@test_broken p_allocated == 0
```

**Comment by charleskawczynski** [comment]
This is an upstream issue: https://github.com/JuliaLang/julia/issues/50554

**Comment by charleskawczynski** [comment]
This is now fixed in Julia 1.10
